render/allocator.c

```diff
@@ -23,21 +24,35 @@ struct wlr_allocator *wlr_allocator_autocreate(struct wlr_backend *backend,
 	uint32_t backend_caps = backend_get_buffer_caps(backend);
 	uint32_t renderer_caps = renderer_get_render_buffer_caps(renderer);
 	int drm_fd = wlr_backend_get_drm_fd(backend);
+	int fd = -1;
+	if (drm_fd != -1) {
+		fd = fcntl(drm_fd, F_DUPFD_CLOEXEC, 0);
+		if (fd < 0) {
+			wlr_log(WLR_ERROR, "fcntl(F_DUPFD_CLOEXEC) failed");
+		}
+	}
 
 	struct wlr_allocator *alloc = NULL;
 	uint32_t gbm_caps = WLR_BUFFER_CAP_DMABUF;
 	if ((backend_caps & gbm_caps) && (renderer_caps & gbm_caps)
-			&& drm_fd != -1) {
+			&& fd != -1) {
 		wlr_log(WLR_DEBUG, "Trying to create gbm allocator");
-		int fd = fcntl(drm_fd, F_DUPFD_CLOEXEC, 0);
-		if (fd < 0) {
-			wlr_log(WLR_ERROR, "fcntl(F_DUPFD_CLOEXEC) failed");
-		} else if ((alloc = wlr_gbm_allocator_create(fd)) != NULL) {
+		if ((alloc = wlr_gbm_allocator_create(fd)) != NULL) {
 			return alloc;
 		}
 		wlr_log(WLR_DEBUG, "Failed to create gbm allocator");
 	}
 
+	uint32_t drm_caps = WLR_BUFFER_CAP_DMABUF | WLR_BUFFER_CAP_DATA_PTR;
+	if ((backend_caps & drm_caps) && (renderer_caps & drm_caps)
+			&& fd != -1) {
```

Comment:
This will likely result in the DRM dumb allocator being picked when running the Pixman renderer with the Wayland or X11 backend. We should probably try the shm allocator first.